pkg/fqdn/identities.go

```diff
@@ -0,0 +1,161 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package fqdn
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/cilium/cilium/pkg/identity"
+	"github.com/cilium/cilium/pkg/labels"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/policy/api"
+	"github.com/cilium/cilium/pkg/trigger"
+)
+
+const (
+	identityAllocationTriggerName     = "fqdn-selector-identity-pre-allocation"
+	identityAllocationTriggerInterval = 1 * time.Second
+)
+
+// newIdentityAllocationQueue creates a new queue to asynchronously allocate
+// and release identities for FQDNSelectors. This is done asynchronously to
+// avoid deadlocks:
+// We must not hold the NameManager lock when calling AllocateLocalIdentity
+// with notifyOwner=true, as the owner is the SelectorCache which can call
+// into NameManager (via RegisterFQDNSelector or UnregisterFQDNSelector),
+// thus creating a cyclic lock dependency which would lead to deadlocks.
+func newIdentityAllocationQueue(allocator IdentityAllocator) *identityQueue {
+	a := newIdentityAllocator(allocator)
+	ctx, cancel := context.WithCancel(context.Background())
+
+	q := &identityQueue{}
+	q.trigger, _ = trigger.NewTrigger(trigger.Parameters{
+		Name:        identityAllocationTriggerName,
+		MinInterval: identityAllocationTriggerInterval,
+		TriggerFunc: func(reasons []string) {
+			for _, item := range q.drain() {
+				if item.allocate {
+					if err := a.allocate(item.selector); err != nil {
+						log.WithError(err).WithField(logfields.Selector, item.selector).
+							Error("Failed to pre-allocate identity")
+					}
+				} else {
+					if err := a.release(ctx, item.selector); err != nil {
+						log.WithError(err).WithField(logfields.Selector, item.selector).
+							Error("Failed to release pre-allocated identity")
+					}
+				}
+			}
+		},
+		ShutdownFunc: cancel,
+	})
+
+	return q
+}
+
+// identityAllocator wraps the real IdentityAllocator and stores the pre-allocated
+// identities for easier release
+type identityAllocator struct {
+	preAllocatedIdentities map[api.FQDNSelector]*identity.Identity
+	backend                IdentityAllocator
+}
+
+func newIdentityAllocator(allocator IdentityAllocator) *identityAllocator {
+	return &identityAllocator{
+		preAllocatedIdentities: make(map[api.FQDNSelector]*identity.Identity),
+		backend:                allocator,
+	}
+}
+
+func (a *identityAllocator) allocate(selector api.FQDNSelector) error {
+	id, ok := a.preAllocatedIdentities[selector]
+	if ok {
+		return errors.New("attempted to pre-allocate already pre-allocated identity")
+	}
+
+	lbls := labels.FromSlice([]labels.Label{selector.IdentityLabel()})
+	id, _, err := a.backend.AllocateLocalIdentity(lbls, true, identity.IdentityUnknown)
```

Comment:
How bad would it be to pass `notifyOwners: false`, then batch all updates to the SelectorCache ourselves?

Reply:
I'll look into it

Reply:
On second thought, let's not do this. We might wind up in a bad race condition where IPs get an identity that is not yet plumbed in.